### PR TITLE
Provide fix-its when overriding a bridged value type with the old reference type

### PR DESCRIPTION
**Pull request by jrose-apple**
This is support for [SE-0069: Mutability and Foundation Value Types](https://github.com/apple/swift-evolution/blob/master/proposals/0069-swift-mutability-for-foundation.md). In cases where someone has overridden a method that takes, e.g. `NSURL`, the override will no longer be valid, because the system
class will now use the value type `URL` instead. If an override's full name matches exactly, the compiler will offer fix-its for any uses of reference types where value types are now preferred. (This must be a direct match; subclasses, including the mutable variants of many Foundation types, will need to be updated by hand.)

One wrinkle here is the use of generics. In Swift 2, Objective-C generics weren't imported at all, so it's unlikely that the overriding method will have the correct generic arguments. Simple migration might insert the "bound" type, but it can't know what specific type might be more appropriate. Therefore, the logic to add the fix-it ignores generic arguments, assuming the parent's type is correct.

rdar://problem/26183575

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by jrose-apple** [comment]
@DougGregor or anyone else who happens by, any suggestions for improved diagnostic text? The "did you mean…" feels a little condescending since we already know the full name matches exactly.

@swift-ci Please test

**Comment by jrose-apple** [comment]
Fixed SourceKit tests. Linux tests previously passed.

@swift-ci Please test OS X platform

**Comment by jrose-apple** [comment]
@swift-ci Please test OS X platform

**Comment by jrose-apple** [comment]
Still hoping for some feedback on the diagnostic before merging.

**Comment by jrose-apple** [comment]
@DougGregor ping

(again)
@swift-ci Please test

**Comment by DougGregor** [comment]
My apologies; in the queue for tomorrow

**Comment by jrose-apple** [comment]
Let's try that again.

@swift-ci Please test

**Comment by gregomni** [comment]
Regarding the diagnostic: Most of the note-type diagnostics are declarative text and point back to what the user should be looking at. Maybe make the diagnostic location at `base` instead of `decl` and just "superclass %0 here".

**Comment by jrose-apple** [comment]
I didn't want to do that because it has a fix-it attached, and it's weird to have the fix-it somewhere far away from the location. (In particular, Xcode will prefer the fix-it location over the diagnostic location, so the text won't match what's shown.)

**Comment by gregomni** [comment]
Ah, okay, didn't realize the fit-it location worked that way. "superclass %0 type is %1" and include the type, then? (I know that it'll also be in the fixit itself, but I think it would be more clear in the message as well.)

**Comment by jrose-apple** [comment]
That's okay, but it doesn't quite match the primary diagnostic, which says that the primary decl "doesn't override". I also don't like that the text doesn't describe what the fix-it does (it's purely declarative).

**Comment by jrose-apple** [comment]
(I appreciate these suggestions a lot!)

**Comment by gregomni** [comment]
Hmm. I think the only way to get more descriptive is to partly repeat the primary diagnostic. Something like "type does not match superclass %0 whose type is %1"? I think shorter and purely declarative for a note is better, though. And thanks for listening to my random suggestions.

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -4786,6 +4786,121 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
       type = fnType->withExtInfo(extInfo);
   }
 
+  static bool fixDeclarationTypes(InFlightDiagnostic &diag, TypeChecker &TC,
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: Please add a doc comment for this function.

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -4786,6 +4786,121 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
...
+      Type normalizedBaseTy = normalizeType(baseTy);
+      const DeclContext *DC = decl->getDeclContext();
+      Optional<Type> maybeBridged =
+          TC.Context.getBridgedToObjC(DC, normalizedBaseTy, &TC);
+
+      // ...and just knowing that it's bridged isn't good enough if we don't
+      // know what it's bridged /to/. Also, don't do this check for trivial
+      // bridging---that doesn't count.
+      Type bridged = maybeBridged.getValueOr(Type());
+      if (!bridged || bridged->isEqual(normalizedBaseTy))
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: I bet we could easily tackle swift_wrapper'd types with this infrastructure.

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -4786,6 +4786,121 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
...
+
+    if (auto *var = dyn_cast<VarDecl>(decl)) {
+      SourceRange typeRange = var->getTypeSourceRangeForDiagnostics();
+      return checkType(var->getType(), base->getType(), typeRange);
+    }
+
+    if (auto *fn = dyn_cast<AbstractFunctionDecl>(decl)) {
+      auto *baseFn = cast<AbstractFunctionDecl>(base);
+      bool fixedAny = false;
+      for_each(*fn->getParameterLists().back(),
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: Are we really really sure these a parameter lists have the same length?

**Comment by DougGregor** [comment]
Minor comments, but this looks great to me.

File: lib/Sema/TypeCheckDecl.cpp (hunk 2 above)
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: That'd be great too, but do we have a good way to get to the wrapped type?

**Comment by jrose-apple** [comment]
I ended up going with Greg's phrasing for the diagnostic (slightly simplified).

@swift-ci Please test

**Comment by jrose-apple** [comment]
Different unrelated Linux failure today! Merging.

**Pull request merged**